### PR TITLE
Fix the issue with the implementation for evaluating the exponential integral

### PR DESCRIPTION
**Pull request by Kvieta1990**
**Description of work.**

The original implementation for evaluating the exponential integral (i.e., the [E1 function](https://en.wikipedia.org/wiki/Exponential_integral)) was wrong, giving weird peak shape. For example, running the following codes in Mantid will produce the weird peak shape as presented below,

```python
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

myFunc = "name=Bk2BkExpConvPV, X0=2949.6002862753003, Intensity=100, Alpha=0.03842, Beta=0.06335, Sigma2=1393.5415922289003, Gamma=6.23432"
ws = CreateSampleWorkspace("Histogram","User Defined", myFunc, BankPixelWidth=1, XUnit='TOF', XMin=2000, XMax=4000, BinWidth=1, NumEvents=1000, NumBanks=1)
```

ws-1

Current PR fixes the issue with the implementation so that E1 function can now be properly evaluated to give correct peak profile. This PR is `critical` for the profile function [`BackToBackExponentialPV`](https://www.mantidproject.org/BackToBackExponentialPV) and [`IkedaCarpenterPV`](https://docs.mantidproject.org/nightly/fitting/fitfunctions/IkedaCarpenterPV.html) (both will use E1 function for calculation) in Mantid to behave accurately.

Since the evaluation for E1 function is changed, the corresponding unit test is also changed.

See the testing section below for the proper output given current implementation in place.

**To test:**

To test the new implementation, we can compare with outputs from several other independent implementations. Using the piece of Python code mentioned above, we can insert some statement just below this line [https://github.com/mantidproject/mantid/blob/4769e2b3fcd198a72f39f970fc11684ee6732626/Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp#L147](https://github.com/mantidproject/mantid/blob/4769e2b3fcd198a72f39f970fc11684ee6732626/Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp#L147) to let Mantid output the calculated profile function. Here attached is the output I got, where x-axis is TOF and y-axis is the calculated profile data, [profile_Mantid.txt](https://github.com/mantidproject/mantid/files/6697158/profile_Mantid.txt).

I compared my implementation with two other external resources. One is a Fortran routine as can be found in the following link, [https://people.sc.fsu.edu/~jburkardt/f_src/special_functions/special_functions.f90](https://people.sc.fsu.edu/~jburkardt/f_src/special_functions/special_functions.f90). The relevant part was extracted to a separate routine, as attached [E1_func.txt](https://github.com/mantidproject/mantid/files/6697179/E1_func.txt). The algorithm in the Fortran routine attached here and the original Mantid implementation is basically identical - it is just that in Mantid there were some issues/typos.

Another independent implementation I used is that in Python Scipy module. Here attached is the full version of the script I created to basically calculate the profile given identical parameters as in the Mantid script attached above - 
[Test_B2BPV.txt](https://github.com/mantidproject/mantid/files/6697219/Test_B2BPV.txt). The profile is output to [profile_python.txt](https://github.com/mantidproject/mantid/files/6697216/profile_python.txt). Comparing the output from Python and my implementation in current PR, we found they give 'identical' (within numerical tolerance) output, as presented below,

E1_Mantid_Python_comparison

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by RichardWaiteSTFC** [comment]
Thanks, I was just about to look into the same problem! ENGIN-X at ISIS have expressed an interest in using this function as well. At present the input parameters include `sigma^2` and `gamma` making it slightly more onerous to tie the parameters such that the Gaussian and Lorentzian components have same FWHM (i.e. `gamma = sqrt(2 ln(2))*sigma`). Can you think of any reason why we can't make the input `sigma` instead of `sigma^2`?

**Comment by Kvieta1990** [comment]
> Thanks, I was just about to look into the same problem! ENGIN-X at ISIS have expressed an interest in using this function as well. At present the input parameters include `sigma^2` and `gamma` making it slightly more onerous to tie the parameters such that the Gaussian and Lorentzian components have same FWHM (i.e. `gamma = sqrt(2 ln(2))*sigma`). Can you think of any reason why we can't make the input `sigma` instead of `sigma^2`?

Thanks for checking! There are several things I want to mention,

1. I think the issue is fixed with my implementation since when I tested locally the profile is correct. I am not sure why the testing is failing.

2. The other thing is, we realized that there is another version of the exponential integral implemented, as can be found here, [https://github.com/mantidproject/mantid/blob/master/Framework/CurveFitting/src/SpecialFunctionHelper.cpp#L24](https://github.com/mantidproject/mantid/blob/master/Framework/CurveFitting/src/SpecialFunctionHelper.cpp#L24). This version is correct. So the intention is to call this version of the exponential integral in Mantid `BackToBackExponentialPV` function.

3. I think the reason why `sigma^2` is used as the input parameter is to follow the same convention that GSAS is using. If you check out GSAS manual, you can see that when GSAS says parameter `sigma0`, it actually means `sigma0^2`.

**Comment by RichardWaiteSTFC** [comment]
> > Thanks, I was just about to look into the same problem! ENGIN-X at ISIS have expressed an interest in using this function as well. At present the input parameters include `sigma^2` and `gamma` making it slightly more onerous to tie the parameters such that the Gaussian and Lorentzian components have same FWHM (i.e. `gamma = sqrt(2 ln(2))*sigma`). Can you think of any reason why we can't make the input `sigma` instead of `sigma^2`?
> 
> Thanks for checking! There are several things I want to mention,
> 
> 1. I think the issue is fixed with my implementation since when I tested locally the profile is correct. I am not sure why the testing is failing.
> 2. The other thing is, we realized that there is another version of the exponential integral implemented, as can be found here, https://github.com/mantidproject/mantid/blob/master/Framework/CurveFitting/src/SpecialFunctionHelper.cpp#L24. This version is correct. So the intention is to call this version of the exponential integral in Mantid `BackToBackExponentialPV` function.
> 3. I think the reason why `sigma^2` is used as the input parameter is to follow the same convention that GSAS is using. If you check out GSAS manual, you can see that when GSAS says parameter `sigma0`, it actually means `sigma0^2`.

Thanks for the above info 
(1 & 2)  So to confirm, you have checked your implementation against the one in `SpecialFunctionHelper.cpp` and both are correct and you just don't want to duplicate it in` Bk2BkExpConvPV`?

(3) Yes thanks I saw that `sigma_i `are already squared in GSAS (and conversely that `gamma_0` etc. are not). I suspect we use it here because you only ever see `Sigma^2` in the function so it stops you evaluating the square every time - but I would argue it is more user friendly to use `Sigma` which can be judged next to the apparent width of the peak (I know the width of the peak is also effected by A, B , gamma). Also `Sigma` is the input in the simpler `BackToBackExponential` and we should probably be consistent (though I think the names are clear enough that people won't be confused).

**Comment by Kvieta1990** [comment]
> > > Thanks, I was just about to look into the same problem! ENGIN-X at ISIS have expressed an interest in using this function as well. At present the input parameters include `sigma^2` and `gamma` making it slightly more onerous to tie the parameters such that the Gaussian and Lorentzian components have same FWHM (i.e. `gamma = sqrt(2 ln(2))*sigma`). Can you think of any reason why we can't make the input `sigma` instead of `sigma^2`?
> > 
> > 
> > Thanks for checking! There are several things I want to mention,
> > 
> > 1. I think the issue is fixed with my implementation since when I tested locally the profile is correct. I am not sure why the testing is failing.
> > 2. The other thing is, we realized that there is another version of the exponential integral implemented, as can be found here, https://github.com/mantidproject/mantid/blob/master/Framework/CurveFitting/src/SpecialFunctionHelper.cpp#L24. This version is correct. So the intention is to call this version of the exponential integral in Mantid `BackToBackExponentialPV` function.
> > 3. I think the reason why `sigma^2` is used as the input parameter is to follow the same convention that GSAS is using. If you check out GSAS manual, you can see that when GSAS says parameter `sigma0`, it actually means `sigma0^2`.
> 
> Thanks for the above info
> (1 & 2) So to confirm, you have checked your implementation against the one in `SpecialFunctionHelper.cpp` and both are correct and you just don't want to duplicate it in` Bk2BkExpConvPV`?
> 
> (3) Yes thanks I saw that `sigma_i `are already squared in GSAS (and conversely that `gamma_0` etc. are not). I suspect we use it here because you only ever see `Sigma^2` in the function so it stops you evaluating the square every time - but I would argue it is more user friendly to use `Sigma` which can be judged next to the apparent width of the peak (I know the width of the peak is also effected by A, B , gamma). Also `Sigma` is the input in the simpler `BackToBackExponential` and we should probably be consistent (though I think the names are clear enough that people won't be confused).

(1 & 2) I check the implementation - both mine and the one in `SpecialFunctionHelper.cpp` - they are both correct. Though, we can stay with my new version of the exponential integral within `Bk2BkExpConvPV`, from futuristic perspective, I think it is beneficial to always use a `central` version of the special function. In another word, we can probably just remove the definition of the exponential integral function within `Bk2BkExpConvPV.cpp`.

3. I totally agree - since when I played with the codes for testing and digging out issues, I always found it is confusing to have `sigma^2` as the input parameter. I will talk to people on our ORNL side to write user stories and potentially work on this issue in a separate PR. That means current PR will be dedicated to solve the profile issue with `Bk2BkExpConvPV` function. Does that sound good to you?

**Comment by RichardWaiteSTFC** [comment]
> (1 & 2) I check the implementation - both mine and the one in `SpecialFunctionHelper.cpp` - they are both correct. Though, we can stay with my new version of the exponential integral within `Bk2BkExpConvPV`, from futuristic perspective, I think it is beneficial to always use a `central` version of the special function. In another word, we can probably just remove the definition of the exponential integral function within `Bk2BkExpConvPV.cpp`.
> 
> (3) I totally agree - since when I played with the codes for testing and digging out issues, I always found it is confusing to have `sigma^2` as the input parameter. I will talk to people on our ORNL side to write user stories and potentially work on this issue in a separate PR. That means current PR will be dedicated to solve the profile issue with `Bk2BkExpConvPV` function. Does that sound good to you?

I agree it's best to keep one central implementation and if you could check with ORNL people about using Sigma that'd be great (and that can be done in a separate PR as you say).
As for the failing test it just looks like the updated value in the assertion wasn't needed - is that the value you got from running the test locally?

**Comment by Kvieta1990** [comment]
> > (1 & 2) I check the implementation - both mine and the one in `SpecialFunctionHelper.cpp` - they are both correct. Though, we can stay with my new version of the exponential integral within `Bk2BkExpConvPV`, from futuristic perspective, I think it is beneficial to always use a `central` version of the special function. In another word, we can probably just remove the definition of the exponential integral function within `Bk2BkExpConvPV.cpp`.
> > (3) I totally agree - since when I played with the codes for testing and digging out issues, I always found it is confusing to have `sigma^2` as the input parameter. I will talk to people on our ORNL side to write user stories and potentially work on this issue in a separate PR. That means current PR will be dedicated to solve the profile issue with `Bk2BkExpConvPV` function. Does that sound good to you?
> 
> I agree it's best to keep one central implementation and if you could check with ORNL people about using Sigma that'd be great (and that can be done in a separate PR as you say).
> As for the failing test it just looks like the updated value in the assertion wasn't needed - is that the value you got from running the test locally?

All sounds good to me!

Concerning the test, that is a bit weird. The updated value in my PR is indeed what I got from testing locally and I am not sure why the testing is still getting the old value with `Bk2BkExpConvPV` already updated. What even weird is, I did the test locally and it seems that no matter how I change the implementation for E1 function (the one in `Bk2BkExpConvPV`), e.g., I multiply the returned `e1` value from the E1 function by 10000 and the testing routine is still getting the old value (which is `y[50] = 2.7983`). Do you have any ideas about this?

**Comment by RichardWaiteSTFC** [comment]
> All sounds good to me!
> 
> Concerning the test, that is a bit weird. The updated value in my PR is indeed what I got from testing locally (by `testing locally`, I mean running the piece of codes that I put on the top of my PR, using identical parameters in the unit test) and I am not sure why the testing is still getting the old value with `Bk2BkExpConvPV` already updated. What even weird is, I did the test locally and it seems that no matter how I change the implementation for E1 function (the one in `Bk2BkExpConvPV`), e.g., I multiply the returned `e1` value from the E1 function by 10000 and the testing routine is still getting the old value (which is `y[50] = 2.7983`). Do you have any ideas about this?

So the test has gamma = 0 so I wouldn't have thought E1 would actually be evaluated? I'm surprised you get a different number locally unless you haven't pushed your latest change to that file?

**Comment by Kvieta1990** [comment]
> > All sounds good to me!
> > Concerning the test, that is a bit weird. The updated value in my PR is indeed what I got from testing locally (by `testing locally`, I mean running the piece of codes that I put on the top of my PR, using identical parameters in the unit test) and I am not sure why the testing is still getting the old value with `Bk2BkExpConvPV` already updated. What even weird is, I did the test locally and it seems that no matter how I change the implementation for E1 function (the one in `Bk2BkExpConvPV`), e.g., I multiply the returned `e1` value from the E1 function by 10000 and the testing routine is still getting the old value (which is `y[50] = 2.7983`). Do you have any ideas about this?
> 
> So the test has gamma = 0 so I wouldn't have thought E1 would actually be evaluated? I'm surprised you get a different number locally unless you haven't pushed your latest change to that file?

Oh, gocha! That is definitely something I want to check! Thank you!

**Comment by Kvieta1990** [comment]
> > All sounds good to me!
> > Concerning the test, that is a bit weird. The updated value in my PR is indeed what I got from testing locally (by `testing locally`, I mean running the piece of codes that I put on the top of my PR, using identical parameters in the unit test) and I am not sure why the testing is still getting the old value with `Bk2BkExpConvPV` already updated. What even weird is, I did the test locally and it seems that no matter how I change the implementation for E1 function (the one in `Bk2BkExpConvPV`), e.g., I multiply the returned `e1` value from the E1 function by 10000 and the testing routine is still getting the old value (which is `y[50] = 2.7983`). Do you have any ideas about this?
> 
> So the test has gamma = 0 so I wouldn't have thought E1 would actually be evaluated? I'm surprised you get a different number locally unless you haven't pushed your latest change to that file?

I know why my local test (with the piece of codes on the top) is showing different value - the X arrays generated in my local testing and the one in the unit test are slightly different so I was not comparing like-to-like. Did not pay much attention to `Gamma` being zero in the test so has been wondering there is something weird beyond my understanding going on. Thank you so much for your help with this!

**Comment by Kvieta1990** [comment]
Hi @RichardWaiteSTFC Is that possible that you can help merging the PR to `master`? Thank you so much for reviewing and the very useful discussion!

**Comment by RichardWaiteSTFC** [comment]
> Hi @RichardWaiteSTFC Is that possible that you can help merging the PR to `master`? Thank you so much for reviewing and the very useful discussion!

No worries, I will review this now.

**Review by RichardWaiteSTFC — COMMENTED** [review]
1. It looks like we only need the last two commits on this branch to be merged? Certainly Tiny change isn't needed/could be squashed to make the history better.
2. Is it possible to add some more tests: specifically one to check that this func with gamma=0 agrees with the `BackToBackExponential` function (should give the same answer) and another to check this function with non-zero gamma.

File: docs/source/release/v6.1.0/diffraction.rst [hunk 1]
```diff
@@ -50,6 +50,7 @@ Improvements
 Bugfixes
 ########
 
+- Fix the issue with peak profile functions :ref:`Bk2BkExpConvPV <func-Bk2BkExpConvPV>`.
```
Comment (RichardWaiteSTFC) [inline docs/source/release/v6.1.0/diffraction.rst:53]: Could the release notes be a bit more specific e.g. 
Correctly calculate exponential integral in :ref:`Bk2BkExpConvPV <func-Bk2BkExpConvPV>`

File: Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp [hunk 2]
```diff
@@ -181,60 +183,13 @@ double Bk2BkExpConvPV::calOmega(double x, double eta, double N, double alpha, do
   if (eta < 1.0E-8) {
     omega2 = 0.0;
   } else {
-    omega2 = 2 * N * eta / M_PI * (imag(exp(p) * E1(p)) + imag(exp(q) * E1(q)));
+    omega2 = 2 * N * eta / M_PI * (imag(exp(p) * exponentialIntegral(p)) + imag(exp(q) * exponentialIntegral(q)));
```
Comment (RichardWaiteSTFC) [inline Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp:186]: I seem to be getting some unusual results for some typical parameters for ENGIN-X - here I compare what I get on this PR with my dev-branch (31691_Bk2BkConvPV_parameters_for_ENGINX) which includes your implementation (before using `SpecialFunctionSupport`) - albeit with some other changes I have made. Could you test again to make sure E1 is the same for both implementations?
image

Comment (RichardWaiteSTFC) [reply Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp:186]: You only see a problem on your branch when `gamma > 0` - so definitely looking like a problem with` E1`

Comment (RichardWaiteSTFC) [reply Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp:186]: Has anyone noticed any problems with the `IkedaCarpenter` function that also uses the `exponentialIntegral` in `SpecialFunctionsSupport`?

Comment (Kvieta1990) [reply Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp:186]: @RichardWaiteSTFC Thanks a lot for the comments. Here is what I have,

1. The first two commits were my fix for the E1 definition within `Bk2BkExpConvPV` but later on we decided to go with the one (`exponentialIntegral`) defined in `SpecialFunctionsSupport` and we ended up with 4 commits where the last two were for the second approach. So basically, if we can make `exponentialIntegral` in `SpecialFunctionsSupport`, we only need the last two commits here plus some other commits to fix the issue with `exponentialIntegral` in `SpecialFunctionsSupport`.

2. I will try to add more tests as you suggested.

3. Will make the release notes more specific.

4. Will definitely check through the two implementations and fix whatever issue with `exponentialIntegral` in `SpecialFunctionsSupport`.

5. I don't think people realized the issue with `exponentialIntegral` in `SpecialFunctionsSupport` before.

> You only see a problem on your branch when gamma > 0 - so definitely looking like a problem with E1

That is definitely true and that really makes the testing with non zero `gamma` necessary.

Comment (Kvieta1990) [reply Framework/CurveFitting/src/Functions/Bk2BkExpConvPV.cpp]: @RichardWaiteSTFC Here is what I got,

1. There is no issue with `exponentialIntegral` in `SpecialFunctionsSupport` and therefore there is no issue with `IkedaCarpenterPV` function either. The problem is with my previous implementation of `exponentialIntegral` in `Bk2BkExpConvPV` - in the implementation for `exponentialIntegral` in `SpecialFunctionsSupport`, it is actually the following that is returned,

    image

    Therefore, the returned value is actually not the E1 function but with an extra multiplicative exponential term. In my previous implementation, I did not realize this and put another multiplicative exponential term in `Bk2BkExpConvPV` and therefore it causes the weird that you see in your example. Unfortunately, since originally we only had the test with gamma being 0 and therefore such a mistake was not caught at all.

2. More tests added.

    2.1. Compare with the output from `BackToBackExponential` when gamma is 0.

    2.2. Test for the situation where gamma is non-zero.

3. The new implementation is actually the same as my very original commit (where the separate version of E1 function defined in `Bk2BkExpConvPV` is used). I have double checked the new implementation by running exactly the same test as what I put in my original PR comments and the results are exactly the same.

4. Release docs changed to reflect a bit more details.

File: Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h [hunk 3]
```diff
@@ -44,5 +45,36 @@ class Bk2BkExpConvPVTest : public CxxTest::TestSuite {
...
+    TS_ASSERT_THROWS_NOTHING(peak_b2bExp.function(x, y_b2bExp));
+    TS_ASSERT_DELTA(y[0], y_b2bExp[0], 1e-4);
+    TS_ASSERT_DELTA(y[50], y_b2bExp[50], 1e-4);
+    TS_ASSERT_DELTA(y[99], y_b2bExp[99], 1e-4);
+
+    // We need a test for non-zero gamma since otherwise the testing
+    // for the exponential integral part is ignored. Given gamma = 1.0,
+    // We still have a peak ~400 but the values are slightly different as
+    // compared to when gamma = 0.
+    peak.setParameter("Gamma", 1.0);
```
Comment (RichardWaiteSTFC) [inline Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h:73]: Great work sorting out the exponential integral, and thanks for adding the tests, but would it be possible to split this test into 3 separate tests? 
1. The original
2. Comparison with `BackToBackExponential` when` Gamma=0`
3. `Gamma > 0`

Comment (Kvieta1990) [reply Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h:73]: @RichardWaiteSTFC Do you mean creating another two separate testing files like `Bk2BkExpConvPVTest1.h` and `Bk2BkExpConvPVTest2.h` to include the other two tests? Sorry I am not seeing why we need to do this, since we are just testing the single profile function?

**Comment by RichardWaiteSTFC** [comment]
I've tested this on ENGIN-X data and it looks good, thanks again for finding and fixing this bug. I'm just waiting on that test to be split up (let me know if you need a hand with this) then I'll approve. I have a PR up for some additional changes, including another smaller bug fix (#31888) - I would appreciate your feedback on it (note I will remove the first commit when this PR is merged).

File: Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h (hunk 3 above)
Comment (RichardWaiteSTFC) [reply Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h:73]: Sorry  - I meant to include three separate tests in the same file - so if e.g. the gamma > 0 test fails we know automatically which part of the function is broken. Let me know if you need a hand with this.

Comment (RichardWaiteSTFC) [reply Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h:73]: For an example see 
https://github.com/mantidproject/mantid/blob/3a31ba7409a9cf020de27f56930ab113e2473654/Framework/CurveFitting/test/Functions/BackToBackExponentialTest.h

Comment (RichardWaiteSTFC) [reply Framework/CurveFitting/test/Functions/Bk2BkExpConvPVTest.h:73]: So instead of having one function inside the test suite
`void test_functionCalculator()`
It'd be good to have at least two with informative names
`void test_function_evaluation_gamma_equal_to_zero()`
`void test_function_evaluation_gamma_greater_than_zero()`

**Comment by Kvieta1990** [comment]
@RichardWaiteSTFC All sounds perfect and I got your point about splitting tests. I think I can do it and commit changes real quick! Thank you for your information!

**Comment by Kvieta1990** [comment]
@RichardWaiteSTFC I split the tests to three separate functions following your comments. Testing it locally did not give any error so I think we should be good to go when the CI checking is finished and passing. Thank you!

**Review by RichardWaiteSTFC — APPROVED** [review]
Test passes locally for me as well so I am approving, thanks again for this!
I've just noticed some more nit-picking things we could do on the test (e.g. TEST_ASSERT_NOTHING when setting the parameters) but I can address this in PR #31888.

**Comment by Kvieta1990** [comment]
> Test passes locally for me as well so I am approving, thanks again for this!
> I've just noticed some more nit-picking things we could do on the test (e.g. TEST_ASSERT_NOTHING when setting the parameters) but I can address this in PR #31888.

Thank you so much for approving and taking care of the extra bits of work!
